### ISSUE TITLE
Server and Collector cannot be installed on the same host

### ISSUE THREAD

**Issue by metmajer** [Dynatrace/Dynatrace-Puppet#23]
The following bug has been submitted by a user: the class declaration `class { 'dynatrace::role::dynatrace_user': }`, which exists in both `dynatrace::role::collector` and `dynatrace::role::server` makes the Puppet run fail with `Error: Duplicate declaration` (when they are invoked in the same manifest):

- https://github.com/Dynatrace/Dynatrace-Puppet/blob/master/manifests/role/collector.pp#L55-L58
- https://github.com/Dynatrace/Dynatrace-Puppet/blob/master/manifests/role/server.pp#L59-L62

**Comment by metmajer** [comment]
A fix is currently undergoing review and will be made available soon.

**Comment by tedfeyler** [comment]
Hey Martin,

I am no longer getting the error for the dynatrace_user, but I am still getting an error on redeclaring File[/var]:

```
Error: Could not retrieve catalog from remote server: Error 400 on SERVER: Duplicate declaration: File[/var] is already declared in file /etc/puppet/modules/dynatrace/manifests/role/server.pp:64; cannot redeclare at /etc/puppet/modules/dynatrace/manifests/role/collector.pp:60 on node va1dynadevsrv03.local
Warning: Not using cache on failed catalog
Error: Could not retrieve catalog; skipping run
```

There is also a conflict between the server.pp and the agents_package.pp:

```
[Error: Could not retrieve catalog from remote server: Error 400 on SERVER: Duplicate declaration: File[/var] is already declared in file /etc/puppet/modules/dynatrace/manifests/role/server.pp:64; cannot redeclare at /etc/puppet/modules/dynatrace/manifests/role/agents_package.pp:42 on node va1dynadevsrv03.local
Warning: Not using cache on failed catalog
Error: Could not retrieve catalog; skipping run](url)

```

While it may not be common to install all three components on a single host, I think that scenario should be supported.  For example, I have a Dynatrace Server and Jenkins Server on the same host so I'd like to have the Agent there as well - which would mean I'd either need a collector or use the embedded collector.

thanks,
ted

**Comment by metmajer** [comment]
Ted, thanks for the feedback. The scenario should definitely be supported and I'll look into it with the `site.pp` you had sent to me on Monday.

**Comment by alfredodeluca** [comment]
Hi Martin. Yes it looks like the Duplicate declaration is in all the class when you try to install 2 services on the same host. 
Alfredo

**Comment by metmajer** [comment]
I have provided a fix and backed it by a test that installs server + collector + agent on a single host. Will have @tedfeyler do a final test run and then push the changes mainline.

**Comment by tedfeyler** [comment]
The latest version you provided works, Martin.  Thanks!!

**Comment by tedfeyler** [comment]
Hey Martin,

I've noticed something strange - and it appears to only happen when you install multiple components on the same host.  You get a circular symbolic link:  you'll find a symbolic link named dynatrace-6.3 in the dynatrace-6.3 directory that links to the dynatrace-6.3 directory (e.g. symbolic link /opt/dynatrace-6.3/dynatrace-6.3 -> /opt/dynatrace-6.3).

Since it only seems to appear when installing multiple components, I thought it appropriate to add it to this thread.

thanks,
ted

**Comment by metmajer** [comment]
The issue has been fixed in the following commits: bace14f16d4e79f31a2de63fabdb75e6ca9d64b2 and 92c35121e831bf4909e5571bb7a46a0c4cd1f698.
Changes have been made available as release v6.3.0-1 on [GitHub](https://github.com/Dynatrace/Dynatrace-Puppet/releases/tag/v6.3.0-1) and on the [PuppetForge](https://forge.puppet.com/dynatrace/dynatrace/6.3.0-1).
